components/json_encoder.rst

```diff
@@ -0,0 +1,767 @@
+The JsonEncoder Component
+=========================
+
+.. warning::
+
+    This component is :doc:`experimental </contributing/code/experimental>` and
+    could be changed at any time without prior notice.
+
+Symfony provides a powerful encoder that is able to encode PHP data
+structures to a JSON stream, and the other way around. That encoder is
+designed to be really efficient, while being able to deal with streams.
+
+The JsonEncoder component should typically be used when working with APIs
+or integrating with third-party APIs.
+
+It can convert an incoming JSON request payload into one or several PHP
+objects that your application can process.
+Then, when sending a response, it can convert the processed PHP objects
+back into a JSON stream for the outgoing response.
+
+Serializer or JsonEncoder?
+--------------------------
+
+When deciding between using of the :doc:`Serializer component <serializer>`
+or the JsonEncoder component, consider the specific needs of your use case.
+
+The Serializer is ideal for scenarios requiring flexibility, such as
+dynamically manipulating object shapes using normalizers and denormalizers,
+or handling complex objects which multiple serialization representation.
+Plus, it allows working with formats beyond JSON (and even with a custom
+format of yours).
+
+On the other hand, the JsonEncoder component is tailored for simple objects
+and offers significant advantages in terms of performance and memory
+efficiency, especially when working with very large JSON.
+Its ability to stream JSON data makes it particularly valuable for handling
+large datasets or dealing with real-time data processing without loading the
+entire JSON into memory.
+
+There is no silver bullet between the Serializer and the JsonEncoder, the
+choice should be guided by the specific requirements of your use case.
+
+Installation
+------------
+
+In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
+install the JsonEncoder component:
+
+.. code-block:: terminal
+
+    $ composer require symfony/json-encoder
+
+.. include:: /components/require_autoload.rst.inc
+
+Configuration
+-------------
+
+The JsonEncoder achieves speed by generating and storing PHP code. To minimize
+code generation during runtime, you can configure it to pre-generate as much
+PHP code as possible during cache warm up. This can be done by specifying
+where the objects to be encoded or decoded are located:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/json_encoder.yaml
+        framework:
+            json_encoder:
+                paths: # where the objects to be encoded/decoded are located
+                    App\Encodable\: '%kernel.project_dir%/src/Encodable/*'
```

Comment:
What do we think about adding this config + the `src/Encodable/` direction to a recipe for `symfony/json-encoder`? Or is there a common use-case of JsonEncoder where you wouldn't want this?

Reply:
Sure, I think that this recipe will be really useful!
About the `Encodable` name, do you think it is clear enough, or should we find another name?

Reply:
I think I was slightly confused here (and only noticed it when you asked about the name). This config option configures where our DTOs live, so we can generate the code at warmup instead of runtime? That might be a bit too application-specific to include in a recipe (would be cool if we could do auto discovery based on a PHP attribute or something, but that's a mechanism we do not yet have for non-services atm).

If I'm understanding this correctly, I would also suggest moving this slightly more down in the documentation. Let's first show people the high level usage, so they understand how the component works, and then suggest they pre-generate the required PHP code.

I guess that will also fix the build error, as `src/Dto/` exists further down in the article (because we create the `Cat` example entity).

(btw, I'm planning on reading the whole article this week, depending on how much work there is in the office :smile: )

Reply:
Yes you're totally right, let's move the "cache generation" section a bit later :slightly_smiling_face: